### ISSUE TITLE
Sobre os testes

### ISSUE THREAD

**Issue by filiperochalopes** [Jogo-da-Biblia/main#38]
A ideia dos testes é verificar se está funcionando o `signin`, `signup`, `criação de pergunta`, `edição de pergunta`, `publicação de pergunta`...(todos os outros que vc fez). Não entendi essa fixture que cria dados fictícios, nem a de cima de exclui. 

![image](https://user-images.githubusercontent.com/34348097/228993871-345c1b85-7ead-437b-95a3-c041b0c372e2.png)

Não vi teste para realizar comentário em pergunta.

**Comment by Ckk3** [comment]
Essa fixture eu tive que criar para declarar os dados padrões para cada teste. Para iniciar cada teste com o mesmo banco de dados, eu faço ela excluir depois adicionar os dados novamente.

**Comment by Ckk3** [comment]
Irei adicionar um teste para realizar comentário em pergunta

**Comment by Ckk3** [comment]
Não entendi o testar signin, signup. Essas queries não estão nos requisitos, você se confundiu?

**Comment by Ckk3** [comment]
Adicionei os testes de comentários. Mantive as fixture de antes.

**Comment by filiperochalopes** [comment]
> Não entendi o testar signin, signup. Essas queries não estão nos requisitos, você se confundiu?

Sim, me confundi. Pq estou com outro projeto tbm agora que tem esses requisitos.

**Comment by filiperochalopes** [comment]
> Essa fixture eu tive que criar para declarar os dados padrões para cada teste. Para iniciar cada teste com o mesmo banco de dados, eu faço ela excluir depois adicionar os dados novamente.

Cada vez que vc criar um dado e exclui no banco de produção, principalmente no quesito de livros e versículos, vc está criando um id novo em cada tabela. O ideal é que vc use outro banco de dados para teste. Pode criar um sqlite de teste ou colocar no docker outra instância do postgres para esse fim. Para não ficar adicionando incrementais no banco de dados de "produção".

**Comment by Ckk3** [comment]
> > Essa fixture eu tive que criar para declarar os dados padrões para cada teste. Para iniciar cada teste com o mesmo banco de dados, eu faço ela excluir depois adicionar os dados novamente.
> 
> Cada vez que vc criar um dado e exclui no banco de produção, principalmente no quesito de livros e versículos, vc está criando um id novo em cada tabela. O ideal é que vc use outro banco de dados para teste. Pode criar um sqlite de teste ou colocar no docker outra instância do postgres para esse fim. Para não ficar adicionando incrementais no banco de dados de "produção".
> 

Esses dados não estão indo pro banco de produção, eles ficam somente dentro do ambiente de teste, o decorator `pytest.mark.db` faz isso

**Comment by Ckk3** [comment]
Retiro oq disse, aparentemente estava indo para o banco de produção. Irei analisar e corrigir

**Comment by Ckk3** [comment]
Erro corrigido, o banco de dados dos testes agora está corretamente isolado.

**Comment by filiperochalopes** [comment]
![image](https://user-images.githubusercontent.com/34348097/229383237-f114469f-9d08-4953-b078-c82553308760.png)

![image](https://user-images.githubusercontent.com/34348097/229383330-53134900-ee55-41a0-82f3-1faff5be940f.png)

Rodando sem sucesso. Certeza que essa é a forma mais simples de popular um banco de dados de teste, colocando cada entrada manualmente quando já temos entradas bases todas em csv? Por exemplo:

![image](https://user-images.githubusercontent.com/34348097/229383279-47777b0b-7d9b-4b7a-afe7-ca95aa1449c7.png)

Já temos a lista de livros em csv.

**Comment by Ckk3** [comment]
> ![image](https://user-images.githubusercontent.com/34348097/229383237-f114469f-9d08-4953-b078-c82553308760.png)
> 
> ![image](https://user-images.githubusercontent.com/34348097/229383330-53134900-ee55-41a0-82f3-1faff5be940f.png)
> 
> Rodando sem sucesso. Certeza que essa é a forma mais simples de popular um banco de dados de teste, colocando cada entrada manualmente quando já temos entradas bases todas em csv? Por exemplo:
> 
> ![image](https://user-images.githubusercontent.com/34348097/229383279-47777b0b-7d9b-4b7a-afe7-ca95aa1449c7.png)
> 
> Já temos a lista de livros em csv.

Você rebuildou a imagem e rodou as migrations novamente? Esse erro de fixture acontece pq não está com o `pytest-django `instalado, ele ja esta nos requirements.txt então é so rebuildar a imagem.

Sobre o db de teste, eu optei por criar dados ficticios só para teste mesmo, a fixture não precisa ser chamada sempre, porém, queria ter o mesmo banco de dados para todos os testes, isso infelizmente não acontecia já que as mudanças que os testes faziam permaneciam para o proximo teste, então para manter a individualidade dos testes eu mantive a criação dos dados em todos eles.

**Comment by Ckk3** [comment]
Se preferir, posso adicionar o banco de dados vigente e criar uma cópia dele para testes, mas sinceramente eu prefiro utilizar uma fixture que gera os campos, me dá uma liberdade e facilidade pra manutenção.

**Comment by filiperochalopes** [comment]
Pronto, tranquilidade. Compreendindo. Verificarei se foi o build.

**Comment by filiperochalopes** [comment]
Instalado, favor remover os warnings
![image](https://user-images.githubusercontent.com/34348097/229962676-b4a2922b-4b4e-4b0e-b39c-f982e9949622.png)

**Comment by Ckk3** [comment]
Corrigido!

**Comment by filiperochalopes** [comment]
Navegando um pouco mais pelos testes percebi que eles não tem todos os casos de uso:

1. Não testa a criação de todos os tipos de Respostas 
2. Não testa a colocação de referências nas perguntas
3. Está sendo feito um hacking com replace para substituir nas strings de requisição, quando tem uma [forma correta de fazer isso](https://docs.graphene-python.org/en/latest/execution/execute/#variables)
4. Uma pergunta: Não entendi pq o id dos usuários começa em 4? Se for pq ele segue a contagem do banco de dados que já existe isso não vai funcionar.
5. Nesse exemplo:
```py
query = pergunta_aleatoria_querie
resultado = client.execute(query, context_value=UsuarioEmContexto(usuario=usuario_admin))
```
É desnecessário atribuir uma variável query, já que ela é somente usada uma vez, em vez disso deveríamos utilizar
```py
resultado = client.execute(pergunta_aleatoria_querie, context_value=UsuarioEmContexto(usuario=usuario_admin))
```
Atribuímos variaveis quando elas serão reaproveitadas
6. Favor reescrever alguns testes com o novo banco de dados temporário que escrevi com dados mais inteligíveis.

**Comment by Ckk3** [comment]
Correções:

- [ ] Teste criação de todos os tipos de Respostas 
- [ ] Teste colocação de referências nas perguntas
- [ ] Substituir a variável query quando nao precisar do replace
- [ ] Utilizar variables do próprio graphene para adicionar valores nas queries
- [ ] Reescrever alguns testes com o novo banco de dados temporário

**Comment by Ckk3** [comment]
"7. Variáveis não utilizadas, todavia persistem no código. Favor realizar limpeza. no exemplo criar_dados_de_teste"
Essa fixtura é chamada pra sempre executar as criação dos dados necessário para os testes, não preciso armazená-la em uma variável para ela executar.

**Comment by Ckk3** [comment]
"2. Não testa a colocação de referências nas perguntas"
Na verdade está testando sim, no teste `test_deve_adicionar_nova_pergunta` o id da referencia é enviado junto com a query

**Comment by Ckk3** [comment]
" Utilizar variables do próprio graphene para adicionar valores nas queries"
Não consegui implementar, provavelmente essa implementação não funciona muito bem com o cliente de teste do graphene. 
Testei tanto com as queries tanto com as mutations e tive o mesmo problema.

**Comment by filiperochalopes** [comment]
> " Utilizar variables do próprio graphene para adicionar valores nas queries" Não consegui implementar, provavelmente essa implementação não funciona muito bem com o cliente de teste do graphene. Testei tanto com as queries tanto com as mutations e tive o mesmo problema.

![image](https://user-images.githubusercontent.com/34348097/232167814-d6cf0a78-1ace-4b02-b5c7-16e9a37fa8e5.png)

Fiz como manda o graphene.

![image](https://user-images.githubusercontent.com/34348097/232167832-c3ed2e2f-3a06-4398-8920-c5fc4d31a6af.png)

Passou tudo direitinho

![image](https://user-images.githubusercontent.com/34348097/232167865-439f8eb0-f36a-4517-9a9a-8138b11119cb.png)

Acho que vc não soube reescrever as queries conforme a [regra GraphQL](https://graphql.org/learn/queries/#variables) . Vou subir esse como exemplo.

**Comment by Ckk3** [comment]
Atualizei todos para esse formato. Pronto pra revisão
